File: LeetCode Questions + Tutorials/EngineeringDigest/Leetcode1721.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ListNode = /** @class */ (function () {
    function ListNode(val, next) {
        if (next === void 0) { next = null; }
        this.val = val;
        this.next = next;
    }
    return ListNode;
}());
function addLast(head, val) {
    var currNode = head;
    while (currNode.next != null) {
        currNode = currNode.next;
    }
    currNode.next = new ListNode(val);
}
function printList(head) {
    var currNode = head;
    var result = "";
    while (currNode != null) {
        result += "".concat(currNode.val, " --> ");
        currNode = currNode.next;
    }
    result += "NULL";
    console.log(result);
}
function swapNodes(head, k) {
    //! Add elements in an array
    var arr = [];
    var currNode = head;
    while (currNode != null) {
        arr.push(currNode.val);
        currNode = currNode.next;
    }
    //! Validate k
    var arrLen = arr.length;
    if (k < 1 || k > arrLen) {
        return head;
    }
    //! Swap elements
    var leftIdx = k - 1;
    var rightIdx = arrLen - k;
    var temp = arr[leftIdx];
    arr[leftIdx] = arr[rightIdx];
    arr[rightIdx] = temp;
    //! Create a new linkedList and return it
    var newHead = new ListNode(arr[0]);
    var tailNode = newHead;
    for (var i = 1; i < arr.length; i++) {
        tailNode.next = new ListNode(arr[i]);
        tailNode = tailNode.next;
    }
    return newHead;
}
;
var head = new ListNode(1);
addLast(head, 2);
addLast(head, 3);
addLast(head, 4);
addLast(head, 5);
console.log("Before swapping: ");
printList(head);
console.log("After swapping: ");
head = swapNodes(head, 2);
printList(head);

File: LeetCode Questions + Tutorials/EngineeringDigest/Leetcode1721.java
import java.util.ArrayList;

class ListNode {
    int data;
    ListNode next;

    ListNode(int data) {
        this.data = data;
        this.next = null;
    }

    ListNode(int data, ListNode next) {
        this.data = data;
        this.next = next;
    }
}

public class Leetcode1721 {
    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        addLast(head, 2);
        addLast(head, 3);
        addLast(head, 4);
        addLast(head, 5);
        System.out.print("Before swapping: ");
        printList(head);
        System.out.print("After swapping: ");
        head = swapNodes(head, 2);
        printList(head);
    }

    private static void addLast(ListNode head, int data) {
        ListNode currNode = head;
        while (currNode.next != null) {
            currNode = currNode.next;
        }
        currNode.next = new ListNode(data);
    }

    private static void printList(ListNode head) {
        ListNode currNode = head;
        while (currNode != null) {
            System.out.print(currNode.data + " --> ");
            currNode = currNode.next;
        }
        System.out.println("Null");
    }

    private static ListNode swapNodes(ListNode head, int k) {
        // * Add List elements into an array.
        ArrayList<Integer> listArr = new ArrayList<>();
        ListNode currNode = head;
        while (currNode != null) {
            listArr.add(currNode.data);
            currNode = currNode.next;
        }
        // ! validate the value of k
        int arrLen = listArr.size();
        if (k < 1 || k > arrLen) {
            return head;
        }

        // ! Swap the elements
        int leftIndex = k - 1;
        int rightIndex = arrLen - k;
        int temp = listArr.get(leftIndex);
        listArr.set(leftIndex, listArr.get(rightIndex));
        listArr.set(rightIndex, temp);

        // ! Rebuild the linked list from the modified ArrayList
        ListNode newHead = new ListNode(listArr.get(0));
        ListNode tail = newHead;
        for (int i = 1; i < listArr.size(); i++) {
            tail.next = new ListNode(listArr.get(i));
            tail = tail.next;
        }
        // ! Return the new head
        return newHead;
    }
}